File: common/src/main/java/app/qwertz/quickjoin/favorites/FavoritesManager.java
package app.qwertz.quickjoin.favorites;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.google.gson.reflect.TypeToken;
import java.io.FileReader;
import java.io.FileWriter;
import java.lang.reflect.Type;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Set;

public class FavoritesManager {

    private static final Gson gson = new GsonBuilder().setPrettyPrinting().create();
    private static final Set<FavoriteEntry> favorites = new LinkedHashSet<>();
    private static Path favoritesFile;

    public static void initialize(Path configDirectory) {
        if (configDirectory == null) return;
        Path quickJoinDir = configDirectory.resolve("QuickJoin");
        try {
            Files.createDirectories(quickJoinDir);
        } catch (Exception ignored) {
            // If we can't create the directory, leave favorites disabled
            favoritesFile = null;
            return;
        }
        favoritesFile = quickJoinDir.resolve("favorites.json");
        load();
    }

    public static boolean isFavorite(String guiName, int buttonId) {
        return favorites.contains(new FavoriteEntry(guiName, buttonId));
    }

    public static boolean toggleFavorite(String guiName, int buttonId) {
        FavoriteEntry entry = new FavoriteEntry(guiName, buttonId);
        boolean added;
        if (favorites.contains(entry)) {
            favorites.remove(entry);
            added = false;
        } else {
            favorites.add(entry);
            added = true;
        }
        save();
        return added;
    }

    public static List<FavoriteEntry> getFavorites() {
        return new ArrayList<>(favorites); // Java 8 compatible
    }
    private static void load() {
        if (favoritesFile == null || !Files.exists(favoritesFile)) return;
        try (FileReader reader = new FileReader(favoritesFile.toFile())) {
            Type type = new TypeToken<List<FavoriteEntry>>() {}.getType();
            List<FavoriteEntry> loaded = gson.fromJson(reader, type);
            favorites.clear();
            if (loaded != null) favorites.addAll(loaded);
        } catch (Exception ignored) {
            favorites.clear(); // malformed file; start clean
        }
    }

    private static void save() {
        if (favoritesFile == null) return;
        try (FileWriter writer = new FileWriter(favoritesFile.toFile(), false)) {
            gson.toJson(favorites, writer);
        } catch (Exception ignored) {
            // ignore save errors to avoid crashing client
        }
    }
}

File: src/main/kotlin/app/qwertz/quickjoin/favorites/FavoritesManager.kt
package app.qwertz.quickjoin.favorites

import com.google.gson.Gson
import com.google.gson.GsonBuilder
import com.google.gson.reflect.TypeToken
import net.minecraft.client.Minecraft
import java.io.File
import java.io.FileReader
import java.io.FileWriter

data class FavoriteEntry(
    val guiName: String,
    val buttonId: Int
)

object FavoritesManager {
    private val gson: Gson = GsonBuilder().setPrettyPrinting().create()
    private val favorites: MutableSet<FavoriteEntry> = linkedSetOf()
    private lateinit var favoritesFile: File

    fun initialize() {
        val mcConfigDir = File(Minecraft.getMinecraft().mcDataDir, "config")
        val quickJoinDir = File(mcConfigDir, "QuickJoin")
        if (!quickJoinDir.exists()) quickJoinDir.mkdirs()
        favoritesFile = File(quickJoinDir, "favorites.json")
        load()
    }

    fun isFavorite(guiName: String, buttonId: Int): Boolean {
        return favorites.contains(FavoriteEntry(guiName, buttonId))
    }

    fun toggleFavorite(guiName: String, buttonId: Int): Boolean {
        val entry = FavoriteEntry(guiName, buttonId)
        val added = if (favorites.contains(entry)) {
            favorites.remove(entry)
            false
        } else {
            favorites.add(entry)
            true
        }
        save()
        return added
    }

    fun getFavorites(): List<FavoriteEntry> {
        return favorites.toList()
    }

    private fun load() {
        if (!::favoritesFile.isInitialized || !favoritesFile.exists()) return
        try {
            FileReader(favoritesFile).use { reader ->
                val type = object : TypeToken<List<FavoriteEntry>>() {}.type
                val loaded: List<FavoriteEntry> = gson.fromJson(reader, type) ?: emptyList()
                favorites.clear()
                favorites.addAll(loaded)
            }
        } catch (_: Exception) {
            // Ignore malformed file; start clean
            favorites.clear()
        }
    }

    private fun save() {
        if (!::favoritesFile.isInitialized) return
        try {
            FileWriter(favoritesFile, false).use { writer ->
                gson.toJson(favorites.toList(), writer)
            }
        } catch (_: Exception) {
            // Ignore save errors to avoid crashing the client
        }
    }
}


